File: Tyche.Android/bet/src/main/java/com/felipearpa/tyche/bet/PoolGamblerBetItemViewState.kt
package com.felipearpa.tyche.bet

import com.felipearpa.tyche.core.emptyString

sealed class PoolGamblerBetItemViewState(open val value: PartialPoolGamblerBetModel) {
    data class Visualization(override val value: PartialPoolGamblerBetModel) :
        PoolGamblerBetItemViewState(value)

    data class Edition(override val value: PartialPoolGamblerBetModel) :
        PoolGamblerBetItemViewState(value)

    companion object
}

fun PoolGamblerBetItemViewState.Companion.emptyVisualization(): PoolGamblerBetItemViewState {
    return PoolGamblerBetItemViewState.Visualization(
        PartialPoolGamblerBetModel(homeTeamBet = emptyString(), awayTeamBet = emptyString())
    )
}

fun PoolGamblerBetItemViewState.copy(value: PartialPoolGamblerBetModel): PoolGamblerBetItemViewState {
    return when (this) {
        is PoolGamblerBetItemViewState.Visualization -> PoolGamblerBetItemViewState.Visualization(
            value
        )

        is PoolGamblerBetItemViewState.Edition -> PoolGamblerBetItemViewState.Edition(value)
    }
}

File: iOS/Bet/Sources/Bet/PoolGamblerBetItemViewState.swift
enum PoolGamblerBetItemViewState {
    case visualization(_ value: PartialPoolGamblerBetModel)
    case edition(_ value: PartialPoolGamblerBetModel)
    
    var value: PartialPoolGamblerBetModel {
        get {
            switch self {
            case .visualization(let value): value
            case .edition(let value): value
            }
        }
        set {
            switch self {
            case .visualization: self = .visualization(newValue)
            case .edition: self = .edition(newValue)
            }
        }
    }
}

extension PoolGamblerBetItemViewState {
    func copy(value: PartialPoolGamblerBetModel) -> PoolGamblerBetItemViewState {
        switch self {
        case .visualization: .visualization(value)
        case .edition: .edition(value)
        }
    }
}

extension PoolGamblerBetItemViewState {
    static func emptyVisualization() -> PoolGamblerBetItemViewState {
        .visualization(PartialPoolGamblerBetModel.empty())
    }
}
